### ISSUE TITLE
(bug): collocated Typescript compiler files not ignored after projen@0.27.35

### ISSUE THREAD

**Issue by njlynch** [projen/projen#1020]
It looks like something changed in the way .gitignore files are generated in #992.

The net impact is that for projects with collocated source + build files (e.g., srcdir === libdir), the `.gitignore` directive for ignoring the `.d.ts` and `.js` files was removed.

I can see that the check is still present in `typescript.ts`:

https://github.com/projen/projen/blob/e66a80f287f643865b0b4ea86081101e96b7f6d7/src/typescript.ts#L268-L275

However, something else must be causing these lines to be removed, as they aren't present in the project once upgraded to ^0.27.35. This even shows up in the Typescript snapshot tests from the above PR:

https://github.com/projen/projen/pull/992/files#diff-11f4dca84bdff503e734bc67efc73c7f54ca3f3ad387a399e61a64316acc5264L20-L21

Example project hit by this:
https://github.com/cdklabs/aws-delivlib/pull/1028/

**Comment by njlynch** [comment]
It looks like `normalizePatterns` is the culprit:

https://github.com/projen/projen/blob/e66a80f287f643865b0b4ea86081101e96b7f6d7/src/ignore-file.ts#L34-L47

With the above:
```ts
const igFile = new IgnoreFile();
igFile.exclude('/lib/**/*.js');
igFile.exclude('/lib/**/*.d.ts');
igFile.include('/lib/');

// igFile._patterns contains only [ '!/lib/' ]
```

**Comment by eladb** [comment]
Assigning to @Chriscbr

**Comment by Chriscbr** [comment]
The small repro you gave above is intended behavior now - it mimics the traditional behavior of .gitignore files, where later statements take precedent over previous ones. To resolve it, we just have to reorder them.

Prior to #992 there was no way to model "keep all of /foo except for /foo/bar" since projen sorted patterns in a way that forced "!/foo" to always get ordered after "/foo/bar".

I've opened a bug fix in #1022. That said I'm still open to feedback if this API makes sense or not in the first place.
